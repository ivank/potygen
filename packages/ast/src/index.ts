import { Parser } from '@ikerin/rd-parse';
import { Grammar } from './grammar';
import { QueryTag, WithTag } from './grammar.types';

export {
  AnyCastTag,
  AnyTypeTag,
  ArrayConstructorTag,
  ArrayIndexRangeTag,
  ArrayIndexTag,
  AsTag,
  BetweenTag,
  BinaryExpressionTag,
  BinaryOperatorTag,
  BooleanTag,
  CaseTag,
  CastableDataTypeTag,
  CastTag,
  CollateTag,
  ColumnsTag,
  ColumnTag,
  CombinationTag,
  ComparationExpressionTag,
  ComparationOperatorTag,
  ConflictConstraintTag,
  ConflictTag,
  ConflictTargetTag,
  ConstantTag,
  CountTag,
  CTETag,
  DataTypeTag,
  DefaultTag,
  DeleteTag,
  DistinctTag,
  DoNothingTag,
  DoUpdateTag,
  ElseTag,
  ExpressionListTag,
  ExpressionTag,
  FromListItemTag,
  FromListTag,
  FromTag,
  FunctionTag,
  GroupByTag,
  HavingTag,
  IdentifierTag,
  InsertTag,
  JoinOnTag,
  JoinTag,
  JoinTypeTag,
  JoinUsingTag,
  LimitTag,
  NullTag,
  NumberTag,
  OffsetTag,
  OperatorExpressionTag,
  OrderByItemTag,
  OrderByTag,
  OrderDirectionTag,
  ParameterTag,
  PgCastTag,
  QuotedNameTag,
  ReturningListItemTag,
  ReturningTag,
  RowTag,
  SelectListItemTag,
  SelectListTag,
  SelectParts,
  SelectTag,
  SetItemTag,
  SetListTag,
  SetMapTag,
  SetTag,
  SqlTag,
  StarIdentifierTag,
  StarTag,
  StringTag,
  TableTag,
  Tag,
  TypeArrayTag,
  TypeTag,
  UnaryExpressionTag,
  UnaryOperatorTag,
  UpdateFromTag,
  UpdateTag,
  UsingTag,
  ValuesListTag,
  ValuesTag,
  WhenTag,
  WhereTag,
  WithTag,
} from './grammar.types';

export {
  isAnyCast,
  isAnyType,
  isArrayConstructor,
  isArrayIndex,
  isArrayIndexRange,
  isAs,
  isBetween,
  isBinaryExpression,
  isBinaryOperator,
  isBoolean,
  isCase,
  isCast,
  isCastableDataType,
  isCollateTag,
  isColumn,
  isColumns,
  isCombination,
  isComparationExpression,
  isComparationOperator,
  isConflict,
  isConflictConstraint,
  isConflictTarget,
  isConstant,
  isCount,
  isCTETag,
  isDataType,
  isDefault,
  isDelete,
  isDistinct,
  isDoNothing,
  isDoUpdate,
  isElse,
  isExpression,
  isExpressionList,
  isFrom,
  isFromList,
  isFromListItem,
  isFunction,
  isGroupBy,
  isHaving,
  isIdentifier,
  isInsertTag,
  isJoin,
  isJoinOn,
  isJoinType,
  isJoinUsing,
  isLimit,
  isNull,
  isNumber,
  isOffset,
  isOperatorExpression,
  isOrderBy,
  isOrderByItem,
  isOrderDirection,
  isParameter,
  isPgCast,
  isQuotedName,
  isReturning,
  isReturningListItem,
  isRow,
  isSelect,
  isSelectList,
  isSelectListItem,
  isSelectTag,
  isSet,
  isSetItem,
  isSetList,
  isSetMap,
  isStar,
  isStarIdentifier,
  isString,
  isTable,
  isType,
  isTypeArray,
  isUnaryExpression,
  isUnaryOperator,
  isUpdate,
  isUpdateFrom,
  isUsing,
  isValues,
  isValuesList,
  isWhen,
  isWhere,
  isWithTag,
  isWrappedExpression,
} from './grammar.guards';

export {
  chunk,
  first,
  groupBy,
  identity,
  initial,
  isDiffBy,
  isEmpty,
  isEqual,
  isNil,
  isObject,
  isUnique,
  isUniqueBy,
  last,
  length,
  orderBy,
  tail,
  toMilliseconds,
} from './util';

export { Grammar } from './grammar';

export const parser = Parser<QueryTag | WithTag>(Grammar);
