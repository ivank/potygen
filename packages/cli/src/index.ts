export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
