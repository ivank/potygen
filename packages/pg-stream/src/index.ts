export { toInternalCursorGenerator } from './internal-cursor-generator';
export { ToEachBatch, ToEachBatchOptions, toEachBatch } from './each-batch';
export { ToReadable, ToReadableOptions, toReadable } from './readable';
export {
  ToAsyncIterator,
  toAsyncIterator,
  ToIteratorOptions,
  ToAsyncBatchIterator,
  toAsyncBatchIterator,
} from './iterator';
