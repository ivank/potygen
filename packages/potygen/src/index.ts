export {
  AstTag,
  AnyCastTag,
  AnyTypeTag,
  ArrayColumnIndexTag,
  ArrayConstructorTag,
  ArrayIndexRangeTag,
  ArrayIndexTag,
  BeginTag,
  BinaryExpressionTag,
  BinaryOperatorTag,
  BitStringTag,
  BooleanTag,
  CaseSimpleTag,
  CaseTag,
  CastableDataTypeTag,
  CastTag,
  CollateTag,
  ColumnsTag,
  ColumnTag,
  CombinationTag,
  CommentTag,
  ComparationArrayInclusionTypeTag,
  ComparationArrayOperatorTag,
  ComparationArrayTypeTag,
  ComparationArrayInclusionTag,
  ComparationArrayTag,
  ExistsTag,
  CompositeAccessTag,
  ConflictConstraintTag,
  ConflictTag,
  ConflictTargetTag,
  ConstantTag,
  CountTag,
  CTENameTag,
  CTETag,
  CTEValuesListTag,
  CTEValuesTag,
  CustomQuotedStringTag,
  DataTypeTag,
  DefaultTag,
  DeleteTag,
  DimensionTag,
  DistinctTag,
  DollarQuotedStringTag,
  DoNothingTag,
  DoUpdateTag,
  ElseTag,
  EmptyLeafSqlTag,
  EmptyLeafTag,
  EscapeStringTag,
  ExpressionListTag,
  ExpressionTag,
  ExtractFieldTag,
  ExtractTag,
  FilterTag,
  FromListItemTag,
  FromListTag,
  FromTag,
  FunctionArgTag,
  FunctionTag,
  GroupByTag,
  HavingTag,
  HexademicalStringTag,
  IdentifierTag,
  InsertTag,
  JoinOnTag,
  JoinTag,
  JoinTypeTag,
  JoinUsingTag,
  LeafSqlTag,
  LimitAllTag,
  LimitTag,
  NamedSelectTag,
  NodeSqlTag,
  NodeTag,
  NullTag,
  NumberTag,
  OffsetTag,
  OperatorExpressionTag,
  OrderByItemTag,
  OrderByTag,
  OrderDirectionTag,
  ParameterTag,
  ParameterPickTag,
  PgCastTag,
  QualifiedIdentifierTag,
  QueryTag,
  QuotedIdentifierTag,
  ReturningListItemTag,
  ReturningTag,
  RollbackTag,
  RowTag,
  RowKeywardTag,
  SavepointTag,
  SelectListItemTag,
  SelectListTag,
  SelectParts,
  SelectTag,
  SetItemTag,
  SetArrayItemTag,
  SetListTag,
  SetMapTag,
  SetTag,
  SqlTag,
  StarIdentifierTag,
  StarTag,
  StringTag,
  TableTag,
  TableWithJoinTag,
  Tag,
  TernaryExpressionTag,
  TernaryOperatorTag,
  TernarySeparatorTag,
  ArrayTypeTag,
  TypeTag,
  UnaryExpressionTag,
  UnaryOperatorTag,
  UnquotedIdentifierTag,
  UpdateFromTag,
  UpdateTag,
  UsingTag,
  ValuesListTag,
  ValuesTag,
  WhenTag,
  WhereTag,
  WithTag,
  WrappedExpressionTag,
  SqlName,
} from './grammar.types';

export {
  isAnyCast,
  isAnyType,
  isArrayColumnIndex,
  isArrayConstructor,
  isArrayIndex,
  isArrayIndexRange,
  isAs,
  isBinaryExpression,
  isBinaryOperator,
  isBitString,
  isBoolean,
  isCase,
  isCaseSimple,
  isCast,
  isCastableDataType,
  isCollate,
  isColumn,
  isColumns,
  isCombination,
  isComment,
  isComparationArrayInclusionType,
  isComparationArrayOperator,
  isComparationArrayType,
  isComparationArrayInclusion,
  isComparationArray,
  isExists,
  isCompositeAccess,
  isConflict,
  isConflictConstraint,
  isConflictTarget,
  isConflictTargetIndex,
  isConstant,
  isCount,
  isCTE,
  isCTEName,
  isCTEValues,
  isCTEValuesList,
  isCustomQuotedString,
  isDataType,
  isDefault,
  isDelete,
  isDistinct,
  isDollarQuotedString,
  isDoNothing,
  isDoUpdate,
  isElse,
  isEmptyLeaf,
  isEscapeString,
  isExpression,
  isExpressionList,
  isFilter,
  isFrom,
  isFromList,
  isFromListItem,
  isFunction,
  isFunctionArg,
  isGroupBy,
  isHaving,
  isHexademicalString,
  isIdentifier,
  isInsert,
  isInteger,
  isJoin,
  isJoinOn,
  isJoinType,
  isJoinUsing,
  isLeaf,
  isLimit,
  isLimitAll,
  isNamedSelect,
  isNode,
  isNull,
  isNumber,
  isOffset,
  isOperatorExpression,
  isOrderBy,
  isOrderByItem,
  isOrderDirection,
  isParameter,
  isParameterPick,
  isPgCast,
  isQualifiedIdentifier,
  isQuotedIdentifier,
  isReturning,
  isReturningListItem,
  isRow,
  isRowKeyward,
  isSelect,
  isSelectList,
  isSelectListItem,
  isSet,
  isSetItem,
  isSetArrayItem,
  isSetList,
  isSetMap,
  isStar,
  isStarIdentifier,
  isString,
  isTable,
  isTableWithJoin,
  isTernaryExpression,
  isArrayType,
  isUnaryExpression,
  isUnaryOperator,
  isUnquotedIdentifier,
  isUpdate,
  isUpdateFrom,
  isUsing,
  isValues,
  isValuesList,
  isWhen,
  isWhere,
  isWith,
  isWrappedExpression,
} from './grammar.guards';

export {
  chunk,
  first,
  groupBy,
  identity,
  initial,
  isDiffBy,
  isEmpty,
  isEqual,
  isNil,
  isObject,
  isUnique,
  isUniqueBy,
  last,
  orderBy,
  tail,
  toMilliseconds,
  range,
  findLastIndex,
} from './util';

export { parser, partialParser } from './grammar';

export {
  TypeString,
  TypeNumber,
  TypeBoolean,
  TypeDate,
  TypeNull,
  TypeJson,
  TypeUnknown,
  TypeLoadCoalesce,
  TypeLoadRecord,
  TypeLoadFunction,
  TypeLoadColumn,
  TypeLoadStar,
  TypeLoadFunctionArgument,
  TypeLoadOperator,
  TypeLoadNamed,
  TypeLoadArray,
  TypeLoadAsArray,
  TypeAny,
  TypeLoadArrayItem,
  TypeContext,
  TypeLiteral,
  BaseTypeLoad,
  BaseType,
  TypeNullable,
  TypeLoadUnion,
  TypeBuffer,
  TypeUnion,
  TypeArray,
  Type,
  TypeOrLoad,
  Result,
  Param,
  ParamPick,
  Source,
  SourceTable,
  SourceQuery,
  QueryInterface,
  TypeLoadObjectLiteral,
  TypeObjectLiteral,
  TypeLoadCompositeAccess,
  TypeComposite,
  TypeLoadOptional,
  TypeOptional,
  TypeLoadColumnCast,
  OperatorVariantPart,
  OperatorVariant,
  TypeName,
} from './query-interface.types';

export {
  isType,
  isTypeString,
  isTypeNumber,
  isTypeBoolean,
  isTypeDate,
  isTypeNull,
  isTypeJson,
  isTypeUnknown,
  isTypeLoadRecord,
  isTypeLoadFunction,
  isTypeLoadColumn,
  isTypeLoadStar,
  isTypeLoadFunctionArgument,
  isTypeLoadOperator,
  isTypeLoadNamed,
  isTypeCoalesce,
  isTypeLoadArray,
  isTypeLoadUnion,
  isTypeArray,
  isTypeNullable,
  isTypeLiteral,
  isTypeUnion,
  isTypeLoadObjectLiteral,
  isTypeAny,
  isTypeLoadArrayItem,
  isTypeLoadAsArray,
  isTypeObjectLiteral,
  isTypeComposite,
  isTypeLoadCompositeAccess,
  isTypeLoadOptional,
  isTypeOptional,
  isTypeLoadColumnCast,
  isSourceTable,
  isSourceQuery,
  isSourceValues,
  isTypeEqual,
  isTypeLoad,
} from './query-interface.guards';

export { SqlInterface, QueryConfig, QuerySource, SqlResult, SqlDatabase, Query, MapQuery, Json } from './sql.types';
export { oneResult, maybeOneResult, mapResult } from './sql.helpers';
export {
  toParams,
  toQueryInterface,
  toConstantBinaryOperatorVariant,
  toAliasedPgType,
  toPgTypeConstant,
  toSources,
  toQueryFrom,
  toQueryResults,
} from './query-interface';
export { sql, toQuery, toQueryConfig } from './sql';
export {
  DatabaseError,
  isDatabaseError,
  PotygenError,
  PotygenDatabaseError,
  PotygenNotFoundError,
  LoadError,
  ParseError,
  ParsedTypescriptFileLoadError,
  ParsedSqlFileLoadError,
} from './errors';

export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
  Path,
  PathItem,
  CompletionEntry,
  InfoContext,
  InfoLoadedQuery,
  QuickInfo,
  Cache,
} from './load.types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
  isLoadedSourceView,
  isLoadedSourceTable,
  isLoadedSourceValues,
  isLoadedSource,
  isLoadedSourceUnknown,
} from './load.guards';

export {
  toLoadedContext,
  loadQueryInterfacesData,
  toLoadedQueryInterface,
  loadData,
  loadAllData,
  extractDataSources,
  filterUnknownLoadedContext,
  throwOnUnknownLoadedContext,
} from './load';

export { LRUCache } from './inspect/cache';
export { closestParent, closestParentPath, toPath } from './inspect/path';
export { completionAtOffset, toInfoContext, quickInfoAtOffset, inspectError } from './inspect/inspect';
