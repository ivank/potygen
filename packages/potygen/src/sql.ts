/**
 * sql.ts
 *
 * The `sql` template literal used to perform all the queries.
 * This holds the user fasing interface
 */

import { parser } from './grammar';
import { AstTag } from './grammar.types';
import { isObject, orderBy } from './util';
import { isDatabaseError, PotygenDatabaseError } from './errors';
import { toParams } from './query-interface';
import { typeUnknown } from './postgres-types-map';
import { Param, TypeOrLoad } from './query-interface.types';
import { Query, QuerySource, SqlDatabase, SqlInterface, QueryConfig } from './sql.types';
import { isType, isTypeLoadNamed, isTypeLoadRecord } from './query-interface.guards';

/**
 * Extract the params from an sql {@link AstTag}.
 */
const toParamsFromAst = (ast: AstTag): Param[] =>
  toParams({ type: typeUnknown, columns: [], cteParams: true })(ast).sort(orderBy((p) => p.start));

const toParamPickType = (type: TypeOrLoad | undefined): string =>
  type
    ? isTypeLoadRecord(type)
      ? type.schema
        ? `::${type.schema}.${type.name}`
        : `::${type.name}`
      : type
      ? isTypeLoadNamed(type)
        ? toParamPickType(type.value)
        : isType(type)
        ? `::${type.postgresType}`
        : ''
      : ''
    : '';

/**
 * Convert a "spread" {@link Param} into params that pg node package will accept
 *
 * $$param -> ($1,$2,$3), ($4,$5,$6)
 */
const toSpreadIndexParam = (param: Param, index: number, values: unknown): string =>
  Array.isArray(values)
    ? values.every(Array.isArray)
      ? values
          .map(
            (value, groupIndex, group) =>
              '(' + value.map((_, itemIndex) => '$' + (index + groupIndex * group.length + itemIndex)).join(',') + ')',
          )
          .join(',')
      : param.pick.length > 0
      ? values
          .map(
            (_, groupIndex) =>
              '(' +
              param.pick.map(
                (pick, itemIndex) =>
                  '$' + (index + groupIndex * param.pick.length + itemIndex) + toParamPickType(pick.castType),
              ) +
              ')',
          )
          .join(',')
      : '(' + values.map((_, itemIndex) => '$' + (index + itemIndex)).join(',') + ')'
    : '()';

/**
 * Figure out what the index will be after a spread
 */
const toSpreadIndex = (param: Param, values: unknown): number =>
  Array.isArray(values)
    ? values.reduce(
        (sum, value) => sum + (Array.isArray(value) ? value.length : isObject(value) ? param.pick.length : 1),
        0,
      )
    : 0;

/**
 * Convert sql string by "spreading" the parameters into a flat list that pg package can interpret.
 * Requies the query to be parsed first and to get params from {@link QueryInterface}
 *
 * ```sql
 * SELECT * FROM id = $param AND age > $age
 * ```
 * Is converted into:
 * ```sql
 * SELECT * FROM id = $1 AND age > $2
 * ```
 *
 */
const convertSql = (params: Param[], sql: string, values: Record<string, unknown>): string =>
  params.reduce<{
    sql: string;
    index: number;
    offset: number;
    indexes: Record<string, number>;
  }>(
    (current, param) => {
      const nameLength = param.end - param.start + 1;
      const reusedIndex = current.indexes[param.name];
      const newIndex = param.spread ? current.index + toSpreadIndex(param, values[param.name]) : current.index + 1;
      const index = reusedIndex ?? newIndex;
      const nextIndex = reusedIndex ? current.index : newIndex;

      const indexParam = param.spread ? toSpreadIndexParam(param, current.index + 1, values[param.name]) : '$' + index;
      const pos = param.start + current.offset;
      return {
        sql: current.sql.slice(0, pos) + indexParam + current.sql.slice(pos + nameLength),
        index: nextIndex,
        indexes: { ...current.indexes, [param.name]: index },
        offset: indexParam.length + current.offset - nameLength,
      };
    },
    { sql, index: 0, offset: 0, indexes: {} },
  ).sql;

/**
 * Convert recieved params objects into the flat array of values that pg package requires
 */
const convertValues = (params: Param[], values: Record<string, unknown>): unknown[] =>
  params.reduce<{ values: unknown[]; indexes: Record<string, boolean> }>(
    (current, param) => {
      const value = values[param.name];
      return {
        values:
          param.name in current.indexes
            ? current.values
            : [
                ...current.values,
                ...(param.spread && Array.isArray(value)
                  ? param.pick.length > 0
                    ? value.flatMap((item) => (isObject(item) ? param.pick.map(({ name }) => item[name]) : item))
                    : value
                  : [value]),
              ],
        indexes: { ...current.indexes, [param.name]: true },
      };
    },
    { values: [], indexes: {} },
  ).values;

/**
 * PG package returns "null" values, this efficiently converts all `null`s into `undefined`s
 */
const nullToUndefinedInPlace = (row: Record<string, unknown>): Record<string, unknown> => {
  for (const key in row) {
    const val = row[key];
    if (val === null) {
      row[key] = undefined;
    } else if (isObject(val) && row.hasOwnProperty(key)) {
      row[key] = nullToUndefinedInPlace(val);
    }
  }
  return row;
};

/**
 * Use the intermediary parsing result {@link QuerySource} to compute a {@link QueryConfig}
 * that can be passed to {@link SqlDatabase}'s query function
 */
const toQueryConfigFromSource = <TSqlInterface extends SqlInterface = SqlInterface>(
  querySource: QuerySource,
  params: TSqlInterface['params'],
): QueryConfig => ({
  text: convertSql(querySource.params, querySource.sql, params as Record<string, unknown>),
  values: convertValues(querySource.params, params as Record<string, unknown>),
});

/**
 * Compute a {@link QueryConfig} from an sql template that can be passed to {@link SqlDatabase}'s query function
 */
export const toQueryConfig = <TSqlInterface extends SqlInterface = SqlInterface>(
  query: Query<TSqlInterface>,
  params: TSqlInterface['params'],
): QueryConfig => toQueryConfigFromSource(query(), params);

export const toQuery = <TSqlInterface extends SqlInterface = SqlInterface>(sql: string): Query<TSqlInterface> => {
  try {
    const { ast } = parser(sql);
    const params = toParamsFromAst(ast);

    return (...args: [db: SqlDatabase, params: TSqlInterface['params']] | []): any => {
      const source = { sql, ast, params };
      if (args.length === 0) {
        return source;
      }

      const query = toQueryConfigFromSource<TSqlInterface>(source, args[1]);

      return args[0]
        .query(query)
        .then((result) => ('rows' in result ? result.rows : result).map(nullToUndefinedInPlace))
        .catch((error) => {
          throw error instanceof Error && isDatabaseError(error) ? new PotygenDatabaseError(error, query) : error;
        });
    };
  } catch (e) {
    console.log(String(e));
    throw e;
  }
};

/**
 * Sql Query. Pass it the {@link SqlInterface} generated with [@potygen/cli](https://github.com/ivank/potygen/tree/main/packages/cli)
 */
export const sql = <TSqlInterface extends SqlInterface = SqlInterface>([
  text,
]: TemplateStringsArray): Query<TSqlInterface> => toQuery(text);
