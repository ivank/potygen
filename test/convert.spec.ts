import { Parser } from '@ikerin/rd-parse';
import { convert } from '../src/convert';
import { SqlGrammar } from '../src/sql.grammar';

const parser = Parser(SqlGrammar);
describe('Convert', () => {
  it.each<[string, string]>([
    ['static fields', `SELECT col1 as "col1", col2 as "myCol2" FROM table1`],
    [
      'join fields',
      `
        SELECT
          t1.col1 as "col1",
          t1.col2 as "myCol2",
          table2.col2 as "other"
        FROM table1 AS t1
        JOIN table2 ON table2.id = table1.id
      `,
    ],
    [
      'join with aliases fields',
      `
        SELECT
          t1.col1 as "col1",
          t1.col2 as "myCol2",
          myT2.col2 as "other",
          myT4.col_9 as "test"
        FROM fit.table1 AS t1
        JOIN fit.table2 as "myT2" ON table2.id = table1.id
        JOIN table3 as "myT4" ON myT2.id = table1.id
      `,
    ],
  ])('Should convert %s sql (%s)', (_, sql) => {
    expect(convert({ params: [], result: [] }, parser(sql))).toMatchSnapshot();
  });
});
