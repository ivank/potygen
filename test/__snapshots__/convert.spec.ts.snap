// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert Should convert join fields sql (
        SELECT
          t1.col1 as "col1",
          t1.col2 as "myCol2",
          table2.col2 as "other"
        FROM table1 AS t1
        JOIN table2 ON table2.id = table1.id
      ) 1`] = `
Object {
  "params": Array [],
  "result": Array [
    Object {
      "name": "col1",
      "type": Object {
        "column": "col1",
        "table": "table1",
      },
    },
    Object {
      "name": "myCol2",
      "type": Object {
        "column": "col2",
        "table": "table1",
      },
    },
    Object {
      "name": "other",
      "type": Object {
        "column": "col2",
        "table": "table2",
      },
    },
  ],
}
`;

exports[`Convert Should convert join with aliases fields sql (
        SELECT
          t1.col1 as "col1",
          t1.col2 as "myCol2",
          myT2.col2 as "other",
          myT4.col_9 as "test"
        FROM fit.table1 AS t1
        JOIN fit.table2 as "myT2" ON table2.id = table1.id
        JOIN table3 as "myT4" ON myT2.id = table1.id
      ) 1`] = `
Object {
  "params": Array [],
  "result": Array [
    Object {
      "name": "col1",
      "type": Object {
        "column": "col1",
        "table": "fit.table1",
      },
    },
    Object {
      "name": "myCol2",
      "type": Object {
        "column": "col2",
        "table": "fit.table1",
      },
    },
    Object {
      "name": "other",
      "type": Object {
        "column": "col2",
        "table": "fit.table2",
      },
    },
    Object {
      "name": "test",
      "type": Object {
        "column": "col_9",
        "table": "table3",
      },
    },
  ],
}
`;

exports[`Convert Should convert static fields sql (SELECT col1 as "col1", col2 as "myCol2" FROM table1) 1`] = `
Object {
  "params": Array [],
  "result": Array [
    Object {
      "name": "col1",
      "type": Object {
        "column": "col1",
        "table": "table1",
      },
    },
    Object {
      "name": "myCol2",
      "type": Object {
        "column": "col2",
        "table": "table1",
      },
    },
  ],
}
`;
